### DISCUSSION TITLE
Suggest files in a folder or create a new file to link

### DISCUSSION THREAD

**Discussion by fallacius** [SilentVoid13/Templater#656] [Help] [answered]
Hello,

I am trying to make a suggester that prompts all the author in "Authors" folder and creates a link to the selected author. If the author doesn't exist, then you would be able to create one.

I have figured out how to get a suggester that prompts all the authors in a folder, but I'm not sure how to add the "create author" option when there is no match. Is there a way I could create a file from a suggester prompt without breaking the flow of the template (eg. other prompts etc.)?

```javascript
<%*
const files = app.vault.getMarkdownFiles();
const authors = new Array();
files.forEach((file) => {
	//console.log("file.path: " + `${file.path}`)
	if (file.path.includes("221 Creators")) {
		authors.push(file.basename)
	};
})
-%>
[[<% tp.system.suggester(authors, authors) %>]]
```

Any help would be appreciated!

**Comment by welpdx** [comment]
> Is there a way I could create a file from a suggester prompt without breaking the flow of the template (eg. other prompts etc.)?

There is no direct way with templater's `tp.system.suggester` implementation.

The [Templater API on `tp.system.suggester`](https://silentvoid13.github.io/Templater/internal-functions/internal-modules/system-module.html#tpsystemsuggestertext_items-string--item-t--string-items-t-throw_on_cancel-boolean--false-placeholder-string---limit-number--undefined) shows what it can and can not do. It can not take the text you input and allow that to be passed out for processing. 

Theoretically, however, if you really really want to retain a one-prompt-only workflow, there is a hack to implement what you want. But I honestly think it is much easier to just add another prompt. 



---

If you want the hacky way, I can make a script on Sunday. (Busy until then). 

~Welp

**Comment by welpdx** [comment, marked as answer]
Ok, it didn't take too long. Also, I realized that even if I were to somehow get the value of the suggester prompt, there is no way I can pass it out. The only way the suggester closes is if you click an option or click outside and exit. The ladder, doesn't allow me to assign the value of what you typed out. 

I think an enhancement issue could be made to ask that the suggester return the value last typed in the suggester input field. 

But anyway, the following implementation works fine. Albeit, there's another prompt that asks you to enter the new author. But it's not too bad at all. 



<details>
           <summary style="font-size:20">Demo</summary>
           <img src="https://i.imgur.com/UjMRzNL.gif" alt="demo" />
</details>



<details><summary>script</summary>
<p>

```js
<%*
const authorFolderName = "221 Creators"
var newAuthorTFolder 
var newAuthorTemplateContent = ""; //template content for new Author //tp.file.find_tfile("template_name")
const files = this.app.vault.getAllLoadedFiles()
const authors = new Array();
files.forEach((file) => {
	if (file.path.includes(authorFolderName) && file.children === undefined) { 
		authors.push(file.basename);
	};
	if (file.name == authorFolderName) {newAuthorTFolder=file;}; // need TFolder to create new file
})

if (authors.length == 0 ){ // authors.pushed nothing in forEach()
	new Notice("Couldn't find Folder: " + authorFolderName);
} else {
	authors.push("Create New Author") // Add last option in suggester
	var author = await tp.system.suggester(authors, authors);
	if (author == null) {
		// suggester exited with no input
		new Notice("Exited Script", 5000)
	} else if (author == "Create New Author") {
		// create author
		if (title = await tp.system.prompt("New Author")) {
		var newfile = await this.app.vault.create(newAuthorTFolder.path+ "/" + title + ".md", newAuthorTemplateContent)
		tR += "[[" + title + "]]";
		} else { new Notice("No Author Input", 5000)}
	} else {
		tR += "[[" + author + "]]";
	}
}
%>
```

Notes: 
- If you don't want to create a new file, just comment out the `this.app.vault.create`. 
- If you want to create a new file, you can add your own template to new file by adding `tp.file.find_tfile("template_name")` to var `newAuthorTemplateContent`

- Unless you have nested folders within `221 Creators`, I suggest not using `file.path.includes("221 Creators")` but maybe `file.parent.name == "221 Creators"`. This way if "221 Creators" was somehow in a folder called "221 Creators2", other files won't be in there. Minor detail though.
</p>
</details>

**Reply by fallacius** [reply]
Hi:

Thank you for the response!

I'm not sure why assigning `newAuthorTemplateContent` as `tp.file.find_tfile("mediaCreator")` returns this error:
`Template parsing error, aborting. 
 The "data" argument must be of type string or an instance of Buffer, TypedArray, or DataView. Received an instance of e`

I'm honestly not sure what this means. When I output the variables, it is just `[object Object]`. (I'm not sure if this helps)

Once again, thank you for the reply!

**Reply by welpdx** [reply]
Opps! My bad. Sorry, please replace `newAuthorTemplateContent` with `await this.app.vault.read(tp.file.find_tfile("mediaCreator"))`

Explaination: 
`tp.file.find_tfile` find the tFile object of your template. 
`this.app.vault.read` will get the content from your template's tFile object.
